smoke-testing/src/test/java/games/strategy/engine/data/GameSaveTest.java

```diff
@@ -0,0 +1,110 @@
+package games.strategy.engine.data;
+
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import games.strategy.engine.ClientFileSystemHelper;
+import games.strategy.engine.framework.GameDataFileUtils;
+import games.strategy.engine.framework.ServerGame;
+import games.strategy.engine.framework.map.file.system.loader.ZippedMapsExtractor;
+import games.strategy.engine.framework.startup.ui.PlayerTypes;
+import games.strategy.engine.framework.startup.ui.panels.main.game.selector.GameSelectorModel;
+import games.strategy.engine.player.Player;
+import games.strategy.net.LocalNoOpMessenger;
+import games.strategy.net.Messengers;
+import games.strategy.net.websocket.ClientNetworkBridge;
+import games.strategy.triplea.settings.ClientSetting;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.triplea.config.product.ProductVersionReader;
+import org.triplea.game.server.HeadlessLaunchAction;
+import org.triplea.injection.Injections;
+import org.triplea.io.ContentDownloader;
+import org.triplea.io.FileUtils;
+
+/**
+ * Checks that no error is encountered when saving a game on several different maps. This test
+ * downloads the maps in question and starts an all-AI game on each of them, before saving.
+ *
+ * <p>Note: A variety of maps are used to ensure different engine features are exercised since
+ * object serialization may run into errors that depend on the state of the object graph.
+ */
+class GameSaveTest {
+
+  @BeforeAll
+  public static void setUp() throws IOException {
+    Injections.init(
+        Injections.builder()
+            .engineVersion(new ProductVersionReader().getVersion())
+            .playerTypes(PlayerTypes.getBuiltInPlayerTypes())
+            .build());
+    ClientSetting.initialize();
+    final Path tempFolder = FileUtils.newTempFolder();
+    FileUtils.writeToFile(tempFolder.resolve(".triplea-root"), "");
+    Files.createDirectory(tempFolder.resolve("assets"));
+    ClientFileSystemHelper.setCodeSourceFolder(tempFolder);
```

Comment:
I presume these lines were needed because it did not work out of the box. The `ClientFileSystemHelper` should be able to find the 'assets' folder without needing this.

Are we compensating for a bug here (which I think https://github.com/triplea-game/triplea/pull/9945 would fix)?

Reply:
This code is creating a temporary directory to be used for the "triplea root", as opposed to using some existing folder. So it needs to set it up to the expected state, which requires the file above and this folder. (Without these, we hit errors.)

I'm not sure it's related to the issue you link.

Reply:
The way TripleA is packaged compared to how it is run locally is somewhat complicated. A lot of places in code (for better or worse) want to know the absolute path of the "triplea root". Those locations then assume a specific folder structure from there. The general idea if that folder structure would be the same in development as it would be in the installation folder. For various reasons this concept has become less & less true over time and more difficult to ensure. Notably the flaws like the 'root' folder depending on the current folder are far more exposed.

Long story long, the '.triplea-root' folder is meant to designate that 'some' folder. That mechanism was created so that the algorithm to find the 'root' folder is to check the current folder and then recurse upwards until we find that specific touch file. That then indicates our 'root' folder.

Why does this matter? There are a lot of assumptions about the root folder. It's a problem to require multiple locations to all set that up correctly, rather it's better if we can do it once & then have all other code properly use the same root folder. Hence, the work here to create a temp folder, create an assets folder (etc..) is duplicating other functionalities that already exist. What this could lead to is having many places that are all setting up the 'root' folder. It is perhaps clear as well that simply by exercising additional functionality there could be more requirements added to this folder setup (which would make the problem go larger in a non-trivial way).

Now that #9945 is merged; I don't know if this setup code is necessary. If it is, then I wonder why the above algorithm for finding the correct 'root' folder is not working.

Reply:
#9945 doesn't change anything. The game logic still requires a directory that has the ".triplea-root" file and an "assets" folder, which is what this is setting up. I rebased just to make sure and confirmed the test fails if I remove any of these three lines.

In terms of such a folder already existing that we can use - I know `game-app/game-headed` has a ".triplea-root" file and I think there's some build step that generates an "assets" folder there. But that build step isn't in the build graph for `:smoke-testing:test`, which I'm sure could be changed via gradle dependencies.

But even that's not enough, because even before that, the error we first get is:
```
No service available of type 'org.triplea.game.ApplicationContext'
org.triplea.util.ServiceNotAvailableException: No service available of type 'org.triplea.game.ApplicationContext'
	at org.triplea.util.Services.lambda$loadAny$0(Services.java:22)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.triplea.util.Services.loadAny(Services.java:22)
	at games.strategy.engine.ClientFileSystemHelper.getCodeSourceFolder(ClientFileSystemHelper.java:64)
	at games.strategy.engine.ClientFileSystemHelper.getRootFolder(ClientFileSystemHelper.java:38)
```

I'm sure we there's a way to set up an application context that will be registered as part of this test and have its location point to the right thing and also to change the gradle build dependencies so that the assets folder will get created as a pre-req before running smoke tests.

But this would result in a much more complicated set up that these three lines of code as part of the test and the necessary changes would be disjoint from the test itself which is not a great design.

Reply:
Needed a few days to come back to this. I am concerned that the three lines in question would become boiler plate and your note about adding a dependency on the task to download assets would be a way to solve. Thank you for posting that 'service' discovery error, that would be hard to resolve and I hope that mechanism can be removed when bots are converted to network relays. I'd agree the better options are therefore not super-feasible.

Reply:
One more thought: If we end up with multiple tests needing this - and therefore this becoming "boilerplate", we could always make a test utility method for this that they all call, so as to keep the logic in just one place. Anyway, I think it's moot for now.